src/scipp/core/reduction.py

```diff
@@ -358,3 +847,111 @@ def any(x: VariableLikeType, dim: Optional[str] = None) -> VariableLikeType:
         return _call_cpp_func(_cpp.any, x)
     else:
         return _call_cpp_func(_cpp.any, x, dim=dim)
+
+
+def _reduce_with_numpy(
+    x: VariableLikeType,
+    *,
+    dim: Dims = None,
+    sc_func: Callable[..., VariableLikeType],
+    np_func: Callable[..., np.ndarray],
+    np_ma_func: Callable[..., np.ndarray],
+    unit_func: Callable[[_cpp.Unit], _cpp.Unit],
+    kwargs: dict[str, Any],
+) -> VariableLikeType:
+    if isinstance(x, _cpp.Dataset):
+        return _cpp.Dataset(
+            {
+                k: _reduce_with_numpy(
+                    v,
+                    dim=dim,
+                    sc_func=sc_func,
+                    np_func=np_func,
+                    np_ma_func=np_ma_func,
+                    unit_func=unit_func,
+                    kwargs=kwargs,
+                )
+                for k, v in x.items()
+            }
+        )
+    if isinstance(x, DataGroup):
+        return data_group_nary(sc_func, x, dim=dim, **kwargs)
```

Comment:
Why `sc_func` and not `_reduce_with_numpy` with the same args used above for dataset?

Reply:
Because it's shorter. I could actually do the same for `Dataset`. It has a tiny performance penalty put that shouldn't matter.